### PR TITLE
Misc 3

### PR DESCRIPTION
**Pull request by tormol**
Seeing #54 reminded me of a bunch of changes I had started on, and motivated me to finish up the non-breaking ones:

* Add `AsciiStr::split()`, `AsciiStr::first()` and `AsciiStr::last()`
* Make `AsciiStr::lines()` iterator double-ended
* Implement `From` conversions between `Cow<AsciiStr>` and `AsciiString` or `&AsciiStr`, and
* Implement creating and `Extend`ing `AsciiString` from `Cow<AsciiStr>`-producing iterators.
* Implement `PartialEq<[u8]>` and `PartialEq<[AsciiChar]>` for `AsciiStr` (in addition to existing `PartialEq<str>`)
* Implement `BorrowMut<AsciiStr>`,  `AsMut<[AsciiChar]>`, `AsRef<[AsciiChar]>` and `AsRef<str>` for `AsciiString`
* Implement `AsRef<AsciiStr>` and `AsMut<AsciiStr>` for `[AsciiChar]`
* Implement `Default` for `AsciiChar`

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -9,7 +9,7 @@ version = "0.9.1"
 
 [dependencies]
 quickcheck = { version = "0.6", optional = true }
-serde = { version = "1.0", optional = true }
+serde = { version = "1.0.25", optional = true }
```
Comment (tomprogrammer) [inline Cargo.toml]: Do we depend on features introduced in 1.0.25? I think it's good practice to keep these dependency specifications as general as possible, to be compatible to binaries locking their dependencies to earlier versions.

For example adding `ascii` to crate `example` shouldn't neccessarily force a update of serde even though `example` already uses an earlier version of serde before 1.0.25.

```suggestion
serde = { version = "1.0", optional = true }
```

File: LICENSE-APACHE [hunk 2]
```diff
@@ -186,8 +186,7 @@
       same "printed page" as the copyright notice for easier
       identification within third-party archives.
 
-   Copyright 2017 Thomas Bahn and contributors
-   Copyright 2014 The Rust Project Developers
+   Copyright [yyyy] [name of copyright owner]
```
Comment (tomprogrammer) [inline LICENSE-APACHE]: Good catch!

File: Cargo.toml (hunk 1 above)
Comment (tormol) [reply Cargo.toml]: Yes, with serde `version = "=1.0.24"` and `version = "=1.0.0"`, I get
```
error[E0407]: method `deserialize_in_place` is not a member of trait `Deserialize`
   --> src/serialization/ascii_string.rs:100:5
    |
100 | /     fn deserialize_in_place<D>(deserializer: D, place: &mut Self) -> Result<(), D::Error>
101 | |     where
102 | |         D: Deserializer<'de>,
103 | |     {
104 | |         deserializer.deserialize_string(AsciiStringInPlaceVisitor(place))
105 | |     }
    | |_____^ not a member of trait `Deserialize`
```
, so `Deserialize::deserialize_in_place()` must have been added in serde 1.0.25.

I've also tested that the minimum version won't affect other crates unless the feature is enabled.

Comment (tomprogrammer) [reply Cargo.toml]: Thanks for the clarification. Obviously that was already true before your PR, so it's a nice fix.
